### PR TITLE
Modifications to group by functions which should help with MySQL 5.7 compatability

### PR DESCRIPTION
**Pull request by SteveIreland**
Sorry about the multiple commits on this.  I meant to consolidate them.

### CODE REVIEW

**Comment by odiea** [comment]
Can this be pulled into a new branch for testing?

**Comment by daN4cat** [comment]
@SteveIreland great, thank you for the contribution.
I'm afraid there are some conflicts to solve but I do hope they are the last ones as I want to release 3.0.1!

File: application/models/Receiving.php [hunk 1]
```diff
@@ -238,28 +238,28 @@ public function create_temp_table(array $inputs)
 			' (INDEX(receiving_date), INDEX(receiving_id))
 			(
 				SELECT 
-					DATE(receiving_time) AS receiving_date,
-					receiving_time,
+					MAX(DATE(receiving_time)) AS receiving_date,
```
Comment (daN4cat) [inline application/models/Receiving.php]: Not sure I get, why MAX function is used here?

File: application/models/Item.php [hunk 2]
```diff
@@ -60,13 +60,61 @@ public function get_found_rows($search, $filters)
...
+
+        $this->db->select('max(item_quantities.item_id) as qty_item_id');
+        $this->db->select('max(item_quantities.location_id) as location_id');
+        $this->db->select('max(item_quantities.quantity) as quantity');
+
+        $this->db->from($this->db->dbprefix('items').' as items');
+        $this->db->join($this->db->dbprefix('suppliers').' as suppliers', 'suppliers.person_id = items.supplier_id', 'left');
+        $this->db->join($this->db->dbprefix('inventory').' as inventory', 'inventory.trans_items = items.item_id');
+
+        $this->db->from($this->db->dbprefix('items').' as items');
```
Comment (daN4cat) [inline application/models/Item.php]: this section looks a copy & paste of the above. Mistake?

File: vendor/codeigniter/framework/system/database/drivers/mysql/mysql_driver.php [hunk 3]
```diff
@@ -281,7 +281,6 @@ protected function _prep_query($sql)
 		{
 			return trim($sql).' WHERE 1=1';
 		}
```
Comment (daN4cat) [inline vendor/codeigniter/framework/system/database/drivers/mysql/mysql_driver.php]: Please revert this change, we don't touch codeigniter original code unless there is a recognised bug stopping us.

File: application/models/Item.php [hunk 4]
```diff
@@ -60,13 +60,61 @@ public function get_found_rows($search, $filters)
 	*/
 	public function search($search, $filters, $rows = 0, $limit_from = 0, $sort = 'items.name', $order = 'asc')
 	{
-		$this->db->from('items');
-		$this->db->join('suppliers', 'suppliers.person_id = items.supplier_id', 'left');
-		$this->db->join('inventory', 'inventory.trans_items = items.item_id');
+
+        $this->db->select('max(items.name) as name');
```
Comment (daN4cat) [inline application/models/Item.php]: I think just using `$this->db->select('items.*, suppliers.*, inventory.*, item_quantities.*')` would do.
Why listing all and then state the AS + max?
It looks an overkill to me.

File: application/models/Item.php [hunk 5]
```diff
@@ -60,13 +60,61 @@ public function get_found_rows($search, $filters)
...
+
+        $this->db->select('max(inventory.trans_id) as trans_id');
+        $this->db->select('max(inventory.trans_items) as trans_items');
+        $this->db->select('max(inventory.trans_user) as trans_user');
+        $this->db->select('max(inventory.trans_date) as trans_date');
+        $this->db->select('max(inventory.trans_comment) as trans_comment');
+        $this->db->select('max(inventory.trans_location) as trans_location');
+        $this->db->select('max(inventory.trans_inventory) as trans_inventory');
+
+        $this->db->select('max(item_quantities.item_id) as qty_item_id');
```
Comment (daN4cat) [inline application/models/Item.php]: This is an optional join, so this could throw errors.

File: application/models/Receiving.php (hunk 1 above)
Comment (jekkos) [inline application/models/Receiving.php]: Group by will now require to use aggregates for every field in the select.

File: application/models/Item.php (hunk 4 above)
Comment (jekkos) [inline application/models/Item.php]: Same thing here.. If we want to use these fields in select we need to use aggregate function

File: application/models/Receiving.php (hunk 1 above)
Comment (daN4cat) [inline application/models/Receiving.php]: Aggregate means `AS` ?

Comment (daN4cat) [inline application/models/Receiving.php]: And what about this `max(` or `MAX(` ?

**Comment by jekkos** [comment]
Thanks for this pull request. Will we stick with full group by sql mode in this case? I think we could also disable this sql mode and stick with how the code is now?

http://stackoverflow.com/questions/23921117/disable-only-full-group-by

**Comment by jekkos** [comment]
We could just add this statement to database.sql

`SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));`

**Comment by jekkos** [comment]
On the other hand if we leave it on we're SQL compliant.

**Comment by daN4cat** [comment]
The `SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));` could go in 3.0.1 and do for the MySQL5.7 support.
For 3.1 we can be SQL compliant do the proper changes.

**Comment by jekkos** [comment]
Ok sounds fine, thanks.

**Comment by odiea** [comment]
Works great. No issue now with MySQL 5.7.

**Comment by jekkos** [comment]
Ok I hoped that it would be persistent for the db configuration itself. Another way to set it is in the database config file itself. This might not be a solution for all in case of shared hosting one does not often have access to db configuration parameters

**Comment by daN4cat** [comment]
@jekkos you mean that adding in the database scripts is not enough/correct?

**Comment by odiea** [comment]
That would be a good idea while this transition is going on. People could continue or even use 5.7 without issues. Once it is resolved that line could be removed.

File: application/models/Item.php (hunk 4 above)
Comment (daN4cat) [inline application/models/Item.php]: OK, I read better about this point as I'm not that expert in SQL stuff and got the point.
No GROUP_BY if there is no aggregate function like SUM, MAX, MIN and similar.

Comment (daN4cat) [inline application/models/Item.php]: The issue here is to understand first why a GROUP_BY is added if there is no aggregate function.
I'll check that part of the code better to understand.

Comment (daN4cat) [inline application/models/Item.php]: In this case we need to solve this problem:

```
		// avoid duplicated entries with same name because of inventory reporting multiple changes on the same item in the same date range
		$this->db->group_by('items.item_id');
```

differently.

File: application/models/Receiving.php (hunk 1 above)
Comment (daN4cat) [inline application/models/Receiving.php]: In Receiving the GROUP BY is probably a copy & paste mistake.
I need to check and remove it, so no need for the MAX(

File: application/models/Item.php (hunk 4 above)
Comment (jekkos) [inline application/models/Item.php]: Think we could try to use DISTINCT, however I'm not sure if it allows one to check for uniqueness of one column only in the select.

File: application/models/Item.php [hunk 6]
```diff
@@ -60,13 +60,61 @@ public function get_found_rows($search, $filters)
...
+        $this->db->select('max(items.name) as name');
+        $this->db->select('max(items.category) as category');
+        $this->db->select('max(items.supplier_id) as supplier_id');
+        $this->db->select('max(items.item_number) as item_number');
+        $this->db->select('max(items.description) as description');
+        $this->db->select('max(items.cost_price) as cost_price');
+        $this->db->select('max(items.unit_price) as unit_price');
+        $this->db->select('max(items.reorder_level) as reorder_level');
+        $this->db->select('max(items.receiving_quantity) as receiving_quantity');
+        $this->db->select('items.item_id as item_id');
```
Comment (daN4cat) [inline application/models/Item.php]: If the group_by is by items.item_id I don't understand how this could work without an aggregate.
It's more a question because I'm 100% sure I got my head around this issue.

Comment (daN4cat) [inline application/models/Item.php]: To test this you need to go to Items and apply a date range to filter the Items selection according to when they where changed.

File: application/models/reports/Specific_discount.php [hunk 7]
```diff
@@ -24,8 +24,10 @@ public function getDataColumns()
 	
 	public function getData(array $inputs)
 	{
-		$this->db->select('sale_id, sale_date, SUM(quantity_purchased) AS items_purchased, customer_name, SUM(subtotal) AS subtotal, SUM(total) AS total, SUM(tax) AS tax, SUM(cost) AS cost, SUM(profit) AS profit, payment_type, comment');
-		$this->db->from('sales_items_temp');
+		$this->db->select('sale_id, MAX(sale_date) AS sales_date, SUM(quantity_purchased) AS items_purchased,
```
Comment (daN4cat) [inline application/models/reports/Specific_discount.php]: This should be AS sale_date and not AS sales_date

File: application/models/reports/Specific_employee.php [hunk 8]
```diff
@@ -24,8 +24,10 @@ public function getDataColumns()
 	
 	public function getData(array $inputs)
 	{
-		$this->db->select('sale_id, sale_date, SUM(quantity_purchased) AS items_purchased, customer_name, SUM(subtotal) AS subtotal, SUM(total) AS total, SUM(tax) AS tax, SUM(cost) AS cost, SUM(profit) AS profit, payment_type, comment');
-		$this->db->from('sales_items_temp');
+		$this->db->select('sale_id, MAX(sale_date) AS sales_date, SUM(quantity_purchased) AS items_purchased,
```
Comment (daN4cat) [inline application/models/reports/Specific_employee.php]: This should be AS sale_date and not AS sales_date
